rfc/rfc-38/rfc-38.md

```diff
@@ -0,0 +1,240 @@
+val translatedOptions = DataSourceWriteOptions.translateSqlOptions(parameters)
+val dfWithoutMetaCols = df.drop(HoodieRecord.HOODIE_META_COLUMNS.asScala:_*)
+
+    if (translatedOptions(OPERATION.key).equals(BOOTSTRAP_OPERATION_OPT_VAL)) {
+      HoodieSparkSqlWriter.bootstrap(sqlContext, mode, translatedOptions, dfWithoutMetaCols)
+    } else {
+      HoodieSparkSqlWriter.write(sqlContext, mode, translatedOptions, dfWithoutMetaCols)
+    }
+    new HoodieEmptyRelation(sqlContext, dfWithoutMetaCols.schema)
+}
+```
+
+Regarding querying, the following method will return a `BaseRelation`（if not provide schema）
+
+```scala
+override def createRelation(sqlContext: SQLContext,
+parameters: Map[String, String]): BaseRelation = {
+createRelation(sqlContext, parameters, null)
+}
+```
+
+For streaming writing and reading, DefaultSource#createSink and DefaultSource#createSink are called respectively.
+In 0.9.0 version , the bulk_insert row mode was introduced to speed up bulk_insert, which implements the `SupportsWrite` v2 api and uses `HoodieDataSourceInternalTable` for writing, 
+right now only bulk_insert operation is supported.
+
+## Implementation
+
+Spark provides a complete V2 api, such as `CatalogPlugin`, `SupportsWrite`, `SupportsRead`, and various pushdown filters,  
+such as `SupportsPushDownFilters`, `SupportsPushDownAggregates`, `SupportsPushDownRequiredColumns`
+
+We would Define the key abstraction of call `HoodieDatasourceTable`, which inherits the `Table`, `SupportsWrite`, `SupportsRead` 
+interfaces to provide writing and reading capabilities.
+
+### Writing Path
+
+Hudi relies heavily on some RDD APIs on the write path, such as the indexing to determine where the record is update or insert, 
+this refactoring work is relatively large under datasource v2 api, so we can firstly falllback to write to v1 
+since Spark provides the V1WriteBuilder interface to bridge the V1 and V2 api, we could implement this in the first stage 
+and perform further reconstruction after there is a better way to deal with indexing and repartitoining with RDD api later. 
+The workflow of writing process of the Datasource V2 can refer to the following figure, 
+where the `DataWriterFactory` is in the Driver and the `DataWriter` that performs writing in the Executor.
+
+![](./1.png)
+
+The writing path code sample is below
+
+```scala
+class HoodieDatasourceTable extends Table with SupportsWrite {
+
+override def name(): String = {
+//
+}
+
+override def schema(): StructType = {
+// get hudi table schema
+}
+
+override def partitioning(): Array[Transform] = {
+// get partitioning of hudi table.
+}
+
+override def capabilities(): Set[TableCapability] = {
+// Set(BATCH_WRITE, BATCH_READ,TRUNCATE,...)
+}
+
+override def newWriteBuilder(info: LogicalWriteInfo): WriteBuilder = {
+// V1WriteBuilder
+}
+
+override def newScanBuilder(CaseInsensitiveStringMap options): ScanBuilder = {
+// HoodieScanBuilder
+}
+
+}
+```
+
+### Querying path
```

Comment:
can you help summarize some high-level benefits and features we'll get by adopting v2 for reading? it'll be nice to have them highlighted in the RFC.

Reply:
done